src/Algebra/Module/Structures.agda

```diff
@@ -0,0 +1,398 @@
+
+    open IsLeftSemimodule isLeftSemimodule
+
+    isSemimodule : IsSemimodule +ᴹ *ₗ (flip *ₗ) 0ᴹ
+    isSemimodule = record
+      { isBisemimodule = record
+        { +ᴹ-isCommutativeMonoid = +ᴹ-isCommutativeMonoid
+        ; isPreleftSemimodule = isPreleftSemimodule
+        ; isPrerightSemimodule = record
+          { *ᵣ-cong = flip *ₗ-cong
+          ; *ᵣ-zeroʳ = *ₗ-zeroˡ
+          ; *ᵣ-distribˡ = *ₗ-distribʳ
+          ; *ᵣ-identityʳ = *ₗ-identityˡ
+          ; *ᵣ-assoc = λ m r s → M-trans (M-sym (*ₗ-assoc s r m))
+                                         (*ₗ-cong (*-comm s r) M-refl)
+          ; *ᵣ-zeroˡ = *ₗ-zeroʳ
+          ; *ᵣ-distribʳ = *ₗ-distribˡ
+          }
+        ; *ₗ-*ᵣ-assoc = λ r m s →
+          M-trans (M-sym (*ₗ-assoc s r m))
+                  (M-trans (*ₗ-cong (*-comm s r) M-refl)
+                           (*ₗ-assoc r s m))
+        }
+      }
+
+    open IsSemimodule isSemimodule public
+      hiding (isLeftSemimodule)
+
+  record IsSemimoduleFromRight (+ᴹ : Op₂ M) (*ᵣ : Opᵣ R M) (0ᴹ : M)
+                              : Set (r ⊔ m ⊔ ℓr ⊔ ℓm) where
+    field
+      isRightSemimodule : IsRightSemimodule semiring +ᴹ *ᵣ 0ᴹ
+
+    open IsRightSemimodule isRightSemimodule
+
+    isSemimodule : IsSemimodule +ᴹ (flip *ᵣ) *ᵣ 0ᴹ
+    isSemimodule = record
+      { isBisemimodule = record
+        { +ᴹ-isCommutativeMonoid = +ᴹ-isCommutativeMonoid
+        ; isPreleftSemimodule = record
+          { *ₗ-cong = flip *ᵣ-cong
+          ; *ₗ-zeroˡ = *ᵣ-zeroʳ
+          ; *ₗ-distribʳ = *ᵣ-distribˡ
+          ; *ₗ-identityˡ = *ᵣ-identityʳ
+          ; *ₗ-assoc = λ r s m → M-trans (*ᵣ-cong M-refl (*-comm r s))
+                                         (M-sym (*ᵣ-assoc m s r))
+          ; *ₗ-zeroʳ = *ᵣ-zeroˡ
+          ; *ₗ-distribˡ = *ᵣ-distribʳ
+          }
+        ; isPrerightSemimodule = isPrerightSemimodule
+        ; *ₗ-*ᵣ-assoc = λ r m s →
+          M-trans (*ᵣ-assoc m r s)
+                  (M-trans (*ᵣ-cong M-refl (*-comm r s))
+                           (M-sym (*ᵣ-assoc m s r)))
+        }
+      }
+
+    open IsSemimodule isSemimodule public
+      hiding (isRightSemimodule)
+
+
+module _ {r ℓr} (ring : Ring r ℓr) where
+  open Ring ring renaming (Carrier to R)
+
+  record IsLeftModule (+ᴹ : Op₂ M) (*ₗ : Opₗ R M) (0ᴹ : M) (-ᴹ : Op₁ M)
+                      : Set (r ⊔ m ⊔ ℓr ⊔ ℓm) where
+    open FP _≈ᴹ_
+    field
+      isLeftSemimodule : IsLeftSemimodule semiring +ᴹ *ₗ 0ᴹ
+      -ᴹ‿cong : Congruent₁ -ᴹ
+      -ᴹ‿inverse : Inverse 0ᴹ -ᴹ +ᴹ
+
+    open IsLeftSemimodule isLeftSemimodule public
+
+    +ᴹ-isAbelianGroup : IsAbelianGroup _≈ᴹ_ +ᴹ 0ᴹ -ᴹ
+    +ᴹ-isAbelianGroup = record
+      { isGroup = record
+        { isMonoid = +ᴹ-isMonoid
+        ; inverse = -ᴹ‿inverse
+        ; ⁻¹-cong = -ᴹ‿cong
+        }
+      ; comm = +ᴹ-comm
+      }
+
+    open IsAbelianGroup +ᴹ-isAbelianGroup public
+      using ()
+      renaming ( isGroup to +ᴹ-isGroup; inverseˡ to -ᴹ‿inverseˡ
```

Comment:
I've tended to go for the style
```agda
  using () renaming
  ( ... to ...
  ; ... to ...
  )
```
as it doesn't end up wasting a load of whitespace.

Reply:
Do you do anything different for long `using` lists?